canvas/biplot.py

```diff
@@ -0,0 +1,499 @@
+            sample_df.groupby(sample_df_metadata[sample_color_category]):
+                ax.plot(np.ravel(group[sample_df_x]),
+                        np.ravel(group[sample_df_y]),
+                        marker=samp_marker, linestyle='', ms=samp_ms,
+                        color=sample_color_dict[name],
+                        label=name, alpha=samp_alpha)
+
+        sample_color_dict = None
+
+    ax2 = ax.twinx()
+
+    if sample_color_category is not None:
+        ax.legend(title=sample_color_category, loc=samp_leg_loc, numpoints=1)
+    else:
+        ax.legend(loc=samp_leg_loc, numpoints=1)
+
+    ax2.set_ylim(ax.get_ylim())
+
+    recs = []
+    feature = []
+    otu_feature_color = []
+    otu_feature_append = []
+    colorVal = []
+
+    if feature_df is not None and feature_df_metadata is None:
+        print('Feature Block 1')
+        if feature_color_category is None:
+            feature_color_category = 'Taxonomy'
+        for name in feature_df.index:
+            otu_feature = name.split(';')[taxa_rank].replace(' ', '')
+            otu_feature_append.append(otu_feature)
+
+        otu_feature_append = sorted(list(set(otu_feature_append)))
+        cNorm = colors.Normalize(vmin=0, vmax=(len(otu_feature_append)-1))
+        scalarMap = cmx.ScalarMappable(norm=cNorm, cmap=feature_colors)
+
+        for index, row in enumerate(otu_feature_append):
+            colorVal.append(scalarMap.to_rgba(index))
+
+        feature_color_dict = dict(zip(otu_feature_append, colorVal))
+        feature_color_dict = OrderedDict(sorted(feature_color_dict.items(),
+                                                key=lambda x: x[0]))
+
+        otu_feature_append = []
+        for name, row in feature_df.iterrows():
+            otu_feature = name.split(';')[taxa_rank].replace(' ', '')
+            ax2.arrow(0, 0,
+                      np.asscalar(row[feature_df_x]),
+                      np.asscalar(row[feature_df_y]),
+                      width=arrow_width, head_width=arrow_head,
+                      alpha=feat_alpha,
+                      color=feature_color_dict[otu_feature])
+
+        for key, value in feature_color_dict.items():
+            recs.append(mpatches.Rectangle((0, 0), 1, 1,
+                                           fc=feature_color_dict[key],
+                                           alpha=feat_alpha))
+            feature.append(key)
+
+        ax2.legend(recs, feature, loc=feat_leg_loc,
+                   title=feature_color_category)
+
+    if feature_df_metadata is not None and feature_color_dict is None:
+        print('Feature Block 2')
+        if feature_color_category is None:
+            feature_color_category = 'Taxonomy'
+
+        for name, group in
+        feature_df.groupby(feature_df_metadata[feature_color_category]):
+            if len(name.split(';')) > 1:
+                otu_feature = name.split(';')[taxa_rank].replace(' ', '')
+                otu_feature_append.append(otu_feature)
```

Comment:
this part of the code is significantly cleaned up and the .split(';') statements are not longer part of the biplot code